contracts/examples/erc1155/src/lib.rs

```diff
@@ -0,0 +1,387 @@
+			"Caller is not approved to transfer tokens from address"
+		);
+
+		if self.get_is_fungible(&type_id) == true {
+			let amount = &value;
+
+			require!(amount > &0, "Must transfer more than 0");
+			let balance = match self.get_balance_mapper(&from).get(&type_id) {
+				Some(b) => b,
+				None => return sc_error!("Address has no tokens of that type"),
+			};
+			require!(amount <= &balance, "Not enough balance for id");
+
+			self.decrease_balance(&from, &type_id, &amount);
+			self.increase_balance(&to, &type_id, &amount);
+		} else {
+			let token_id = &value;
+
+			require!(
+				self.is_valid_token_id(&type_id, &token_id),
+				"Token type-id pair is not valid"
+			);
+			require!(
+				self.get_token_owner(&type_id, &token_id) == from,
+				"_from_ is not the owner of the token"
+			);
+
+			let amount = BigUint::from(1u32);
+			self.decrease_balance(&from, &type_id, &amount);
+			self.increase_balance(&to, &type_id, &amount);
+
+			self.set_token_owner(&type_id, token_id, &to);
+		}
+
+		// self.transfer_single_event(&caller, &from, &to, &id, &amount);
+
+		if self.is_smart_contract_address(&to) {
+			// TODO: async-call
+		}
+
+		Ok(())
+	}
+
+	/// `value` is amount for fungible, token_id for non-fungible
+	#[endpoint(safeBatchTransferFrom)]
+	fn safe_batch_transfer_from(
+		&self,
+		from: Address,
+		to: Address,
+		type_ids: &[BigUint],
+		values: &[BigUint],
+		_data: &[u8],
+	) -> SCResult<()> {
+		let caller = self.get_caller();
+
+		require!(
+			caller == from || self.get_is_approved(&caller, &from),
+			"Caller is not approved to transfer tokens from address"
+		);
+		require!(to != Address::zero(), "Can't transfer to address zero");
+		require!(
+			type_ids.len() == values.len(),
+			"Id and value lenghts do not match"
+		);
+
+		// storage edits are rolled back in case of SCError, 
+		// so the reverting is handled automatically if one of the transfers fails
+		for i in 0..type_ids.len() {
+			let type_id = &type_ids[i];
+
+			if self.get_is_fungible(type_id) == true {
+				let amount = &values[i];
+
+				let balance = match self.get_balance_mapper(&from).get(type_id) {
+					Some(b) => b,
+					None => return sc_error!("Address has no tokens of that type"),
+				};
+
+				require!(amount > &0, "Must transfer more than 0");
+				require!(amount <= &balance, "Not enough balance for id");
+
+				self.decrease_balance(&from, &type_id, &amount);
+				self.increase_balance(&to, &type_id, &amount);
+			} else {
+				let token_id = &values[i];
+
+				require!(
+					self.get_token_owner(&type_id, &token_id) == from,
+					"_from_ is not the owner of the token"
+				);
+
+				let amount = BigUint::from(1u32);
+
+				self.decrease_balance(&from, &type_id, &amount);
+				self.increase_balance(&to, &type_id, &amount);
+
+				self.set_token_owner(&type_id, &token_id, &to);
+			}
+		}
+
+		// self.transfer_batch_event(&caller, &from, &to, ids, amounts);
+
+		if self.is_smart_contract_address(&to) {
+			// TODO: async-call
```

Comment:
in case of smart contracts you need to set an async call and have the state for increase / decrease in a pending state - thus if the async call fails you should revert those.